docs/cxx-proposal/cxx.rst

```diff
@@ -0,0 +1,469 @@
+where ``P`` is itself a serializable property type.  Accordingly, one is free
+to declare a property of type ``std::vector<std::map<std::string, std::list<TimeRange>>>`` and it will
+serialize and deserialize properly.  However, such a type might be hard to reflect/bind in
+a Python or Swift bridge.  Our current implementation however bridges one-level deep types
+such as ``std::vector<RationalTime>`` or ``std::map<std::string, double>`` to Python (and later Swift)
+quite easily and idiomatically.
+
+Finally, one can declare lists and dictionaries for schema objects, in as strongly typed
+fashion as required.  That is, a property might be a list of schema objects of any type,
+or the property might specify a particular derived class the schema object must satisfy.
+Again, this is taken care of automatically: ::
+
+  class DerivedSchema : public SerializableObject {
+     ...
+  private:
+     std::vector<MediaReference*> _extra_references;   // (don't actually do this)
+  };
+
+In this case, the derived schema could choose to store extra media references.  The reading/writing
+code would simply call ::
+
+   reader.read("extra_references", &_extra_references)
+
+and ::
+
+    writer.write("extra_references", _extra_references)
+
+to read/write the property.
+
+.. Note::
+   The comment "don't actually do this" will be explained in the next section; the
+   actual type of this property needs to be slightly different.
+   The code for reading/writing the property however is correct.
+   
+Object Graphs and Serialization
++++++++++++++++++++++++++++++++
+
+The current Python implementation assumes that no schema object is referenced
+more than once, when it comes to serialization and deserialization.  Specifically, the
+object "graph" is assumed to implicitly be a tree, although this is not always enforced.
+For example, the current Python implementation has this bug: ::
+
+  clip1 = otio.schema.Clip("clip1")
+  clip2 = otio.schema.Clip("clip2")
+  ext_ref = otio.schema.ExternalReference("/path/someFile.mov")
+  clip1.media_reference = ext_ref
+  clip2.media_reference = ext_ref
+
+As written, modifying ``ext_ref`` modifies the external media reference data for
+both ``clip1`` and ``clip2``.  However, if one serializes and then deserializes this
+data, the serialized data replicates the external references.  Thus, upon reading
+back this object graph, the new clips no longer share the same media reference.
+
+The C++ implementation for serialization will not have this limitation.
+That means that the object structure need no longer be a tree; it doesn't, strictly speaking, even need
+to be a DAG: ::
+
+   Clip* clip1 = new Clip();
+   Clip* clip2 = new Clip();
+
+   clip1->metadata()["other_clip"] = clip2;
+   clip2->metadata()["other_clip"] = clip1;
+
+will work just fine: writing/reading or simply cloning ``clip1`` would yield
+a new ``clip1`` that pointed to a new ``clip2`` and vice versa.
```

Comment:
We should have a larger discussion about either enforcing the tree structure at the schema level, or explicitly allowing non-tree structures.

We suspect that some future schemas/features will want to use this (e.g. Effects that use multiple clips as sources, or linked audio & video clips on two different tracks) but so far we have been very happy with the simplicity of the tree structure as an aid for making OTIO useful and easy to learn.

In any case, if/when we choose to allow for non-tree structures, we should be explicit and intentional about it.